packages/firestore/exp/test/shim.ts

```diff
@@ -0,0 +1,768 @@
+  readonly path = this._delegate.path;
+
+  get parent(): DocumentReference<legacy.DocumentData> | null {
+    const docRef = parent(this._delegate);
+    return docRef ? new DocumentReference<legacy.DocumentData>(docRef) : null;
+  }
+
+  doc(documentPath?: string): DocumentReference<T> {
+    if (documentPath !== undefined) {
+      return new DocumentReference<T>(doc(this._delegate, documentPath));
+    } else {
+      return new DocumentReference<T>(doc(this._delegate));
+    }
+  }
+
+  add(data: T): Promise<DocumentReference<T>> {
+    return addDoc(this._delegate, unwrap(data)).then(
+      docRef => new DocumentReference(docRef)
+    );
+  }
+
+  isEqual(other: CollectionReference<T>): boolean {
+    return refEqual(this._delegate, other._delegate);
+  }
+
+  withConverter<U>(
+    converter: legacy.FirestoreDataConverter<U>
+  ): CollectionReference<U> {
+    return new CollectionReference<U>(
+      this._delegate.withConverter(
+        converter as UntypedFirestoreDataConverter<U>
+      )
+    );
+  }
+}
+
+export class FieldValue implements legacy.FieldValue {
+  constructor(readonly _delegate: exp.FieldValue) {}
+
+  static serverTimestamp(): FieldValue {
+    return new FieldValue(serverTimestamp());
+  }
+
+  static delete(): FieldValue {
+    return new FieldValue(deleteField());
+  }
+
+  static arrayUnion(...elements: any[]): FieldValue {
+    return new FieldValue(arrayUnion(...unwrap(elements)));
+  }
+
+  static arrayRemove(...elements: any[]): FieldValue {
+    return new FieldValue(arrayRemove(...unwrap(elements)));
+  }
+
+  static increment(n: number): FieldValue {
+    return new FieldValue(increment(n));
+  }
+
+  isEqual(other: FieldValue): boolean {
+    return this._delegate.isEqual(other._delegate);
+  }
+}
+
+export class FieldPath implements legacy.FieldPath {
+  private fieldNames: string[];
+
+  constructor(...fieldNames: string[]) {
+    this.fieldNames = fieldNames;
+  }
+
+  get _delegate(): FieldPathExp {
+    return new FieldPathExp(...this.fieldNames);
+  }
+
+  static documentId(): FieldPath {
+    return new FieldPath('__name__');
+  }
+
+  isEqual(other: FieldPath): boolean {
+    throw new Error('isEqual() is not supported in shim');
+  }
+}
+
+/**
+ * Takes document data that uses the firestore-exp API types and replaces them
+ * with the API types defined in this shim.
+ */
+function wrap(value: any): any {
+  if (Array.isArray(value)) {
+    return value.map(v => wrap(v));
+  } else if (value instanceof FieldPathExp) {
+    return new FieldPath(...value._internalPath.toArray());
+  } else if (value instanceof DocumentReferenceExp) {
+    return new DocumentReference(value);
+  } else if (isPlainObject(value)) {
+    const obj: any = {};
+    for (const key in value) {
+      if (value.hasOwnProperty(key)) {
+        obj[key] = wrap(value[key]);
+      }
+    }
+    return obj;
+  } else {
+    return value;
+  }
+}
+
+/**
+ * Takes user data that uses API types from this shim and replaces them
```

Comment:
ultra-uber-nit (so that I can say I left a comment): Move "with" to line above.

Reply:
Done

Reply:
(Sorry - I just realized I forgot to do this :/)